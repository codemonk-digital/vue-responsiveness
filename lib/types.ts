export type VueResponsivenessBreakpoints = Record<string, number | null>

export type VueResponsivenessMatches = Record<
  string,
  { min: boolean; max: boolean; only: boolean }
> & {
  current: string
  isMin(interval: string): boolean
  isMax(interval: string): boolean
  isOnly(interval: string): boolean
}
