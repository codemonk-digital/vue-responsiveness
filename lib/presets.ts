import type { VueResponsivenessBreakpoints } from './types'

export type PresetName =
  | 'Ant_Design'
  | 'Bootstrap_3'
  | 'Bootstrap_4'
  | 'Bootstrap_5'
  | 'Bulma'
  | 'Chakra'
  | 'Foundation'
  | 'Ionic'
  | 'Master_CSS'
  | 'Material_Design'
  | 'Materialize'
  | 'Material_UI'
  | 'Quasar'
  | 'Semantic_UI'
  | 'Skeleton'
  | 'Tailwind_CSS'
  | 'Vuetify'
  | 'Windi_CSS'

export const Presets: Record<PresetName, VueResponsivenessBreakpoints> = {
  Ant_Design: {
    xs: 0,
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200,
    xxl: 1600
  },
  Bootstrap_3: {
    xs: 0,
    sm: 576,
    lg: 992
  },
  Bootstrap_4: {
    xs: 0,
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200
  },
  Bootstrap_5: {
    xs: 0,
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200,
    xxl: 1400
  },
  Bulma: {
    mobile: 0,
    tablet: 769,
    desktop: 1024,
    widescreen: 1216,
    fullhd: 1408
  },
  Chakra: {
    xs: 0,
    sm: 480,
    md: 768,
    lg: 992,
    xl: 1280,
    '2xl': 1472
  },
  Foundation: {
    small: 0,
    medium: 640,
    large: 1024,
    xlarge: 1200,
    xxlarge: 1440
  },
  Ionic: {
    xs: 0,
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200
  },
  Master_CSS: {
    '3xs': 360,
    '2xs': 480,
    xs: 600,
    sm: 768,
    md: 1024,
    lg: 1280,
    xl: 1440,
    '2xl': 1600,
    '3xl': 1920,
    '4xl': 2560
  },
  Material_Design: {
    xs: 0,
    sm_8: 600,
    sm_12: 905,
    md: 1240,
    lg: 1440
  },
  Materialize: {
    s: 0,
    m: 600,
    l: 992,
    xl: 1200
  },
  Material_UI: {
    xs: 0,
    sm: 600,
    md: 960,
    lg: 1260,
    xl: 1920
  },
  Quasar: {
    xs: 0,
    sm: 600,
    md: 1024,
    lg: 1440,
    xl: 1920
  },
  Semantic_UI: {
    mobile: 0,
    tablet: 768,
    small_monitor: 992,
    large_monitor: 1200
  },
  Skeleton: {
    mobile: 0,
    phablet: 400,
    tablet: 550,
    desktop: 750,
    desktop_hd: 1000,
    desktop_hd_lg: 1200
  },
  Tailwind_CSS: {
    xs: 0,
    sm: 640,
    md: 768,
    lg: 1024,
    xl: 1280,
    '2xl': 1536
  },
  Vuetify: {
    xs: 0,
    sm: 600,
    md: 960,
    lg: 1280,
    xl: 1920
  },
  Windi_CSS: {
    xs: 0,
    sm: 640,
    md: 768,
    lg: 1024,
    xl: 1280,
    '2xl': 1536
  }
}
