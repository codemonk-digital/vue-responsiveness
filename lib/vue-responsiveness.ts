import type { VueResponsivenessBreakpoints, VueResponsivenessMatches } from './'
import type { App } from 'vue'
import { Presets } from './'
import { reactive, computed } from 'vue'

let matches: VueResponsivenessMatches

export const VueResponsiveness = {
  install(
    app: App,
    breakpoints: VueResponsivenessBreakpoints = Presets.Bootstrap_5
  ): void {
    const intervals = Object.entries(breakpoints)
      .sort(([, a], [, b]) => Number(a) - Number(b))
      .reduce(
        (out, [key, min], i, arr) => {
          out[key] = {
            min: min ? `(min-width: ${min}px)` : '',
            max: arr[i + 1]?.[1]
              ? `(max-width: ${(arr[i + 1][1] as number) - 0.1}px)`
              : ''
          }
          return out
        },
        {} as Record<string, { min: string; max: string }>
      )
    matches = reactive({
      isMin: computed(() => (key: string) => matches[key]?.min || false),
      isMax: computed(() => (key: string) => matches[key]?.max || false),
      isOnly: computed(() => (key: string) => matches[key]?.only || false),
      current: computed(
        () => Object.keys(intervals).find((key) => matches[key].only) || ''
      ),
      ...Object.keys(intervals).reduce((acc, key) => {
        acc[key] = { min: false, max: false, only: false }
        return acc
      }, {} as Partial<VueResponsivenessMatches>)
    }) as VueResponsivenessMatches

    if (typeof window !== 'undefined') {
      Object.entries(intervals).forEach(([interval, mediaQueries]) => {
        const queryLists = {
          min: window.matchMedia(mediaQueries.min),
          max: window.matchMedia(mediaQueries.max)
        }
        Object.entries(queryLists).forEach(([key, mediaQueryList]) => {
          const listener = ({
            matches: val
          }: MediaQueryListEventMap['change']) => {
            const { min, max } = {
              ...matches[interval],
              [key]: val
            } as { min: boolean; max: boolean }
            matches[interval] = { min, max, only: min && max }
          }
          mediaQueryList.addEventListener('change', listener)
          listener(
            mediaQueryList as unknown as MediaQueryListEventMap['change']
          )
        })
      })
    }

    app.config.globalProperties.$matches = matches
  }
}

export const useMatches = () => matches
