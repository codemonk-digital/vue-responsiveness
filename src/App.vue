<template>
  <div>
    <pre v-text="JSON.stringify($matches, null, 2)" />
    <h1>
      Current interval:
      <span>{{ $matches.current }}</span>
    </h1>
    <h3 v-if="$matches.isMax('sm')"><code>sm</code> and below</h3>
    <h3 v-if="$matches.isOnly('md')"><code>md</code> only</h3>
    <h3 v-if="$matches.isMin('lg')"><code>lg</code> and above</h3>
  </div>
</template>
