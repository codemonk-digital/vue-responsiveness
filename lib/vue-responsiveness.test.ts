import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import { defineComponent } from 'vue'
import App from '../src/App.vue'
import { VueResponsiveness, useMatches } from './vue-responsiveness'
import { type VueResponsivenessBreakpoints } from './types'
import { Presets } from './presets'
import { vi } from 'vitest'

describe('vue-responsiveness', () => {
  const bs5 = Object.assign(
    {},
    {
      current: 'xs'
    },
    ...Object.keys(Presets.Bootstrap_5).map((key) => ({
      [key]: {
        max: true,
        min: true,
        only: true
      }
    }))
  )
  const render = (breakpoints?: VueResponsivenessBreakpoints) =>
    mount(App, {
      global: {
        plugins: [[VueResponsiveness, breakpoints]]
      }
    })

  it('should work', () => {
    expect(render().element).toMatchSnapshot()
  })

  it('should work with custom breakpoints', () => {
    expect(render({ min: null, max: 1000 }).element).toMatchSnapshot()
  })

  it('should reverse unordered values', () => {
    expect(render({ max: 1000, min: null }).element).toMatchSnapshot()
  })

  it('should match $matches', () => {
    const { isMax, isMin, isOnly, ...rest } = render().vm.$matches
    expect(rest).toEqual(bs5)
    expect(isMax('sm')).toBe(true)
    expect(isMin('sm')).toBe(true)
    expect(isOnly('sm')).toBe(true)
  })

  it('useMatches() should work', () => {
    const App = defineComponent({
      setup: () => ({ matches: useMatches() }),
      template: '<div />'
    })
    const { vm } = mount(App, {
      global: {
        plugins: [[VueResponsiveness]]
      }
    })
    const { isMax, isMin, isOnly, ...rest } = vm.matches
    expect(rest).toEqual(bs5)
    expect(isMax('sm')).toBe(true)
    expect(isMin('sm')).toBe(true)
    expect(isOnly('sm')).toBe(true)
  })

  it('should return window.matchMedia matches', () => {
    window.matchMedia = vi.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: vi.fn(), // deprecated
      removeListener: vi.fn(), // deprecated
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      dispatchEvent: vi.fn()
    }))
    const { isMax, isMin, isOnly, current } = render().vm.$matches
    expect(isMax('sm')).toBe(false)
    expect(isMin('sm')).toBe(false)
    expect(isOnly('sm')).toBe(false)
    expect(current).toBe('')
  })
})
