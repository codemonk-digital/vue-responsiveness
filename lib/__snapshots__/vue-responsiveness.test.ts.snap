// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vue-responsiveness > should reverse unordered values 1`] = `
<div>
  <pre>
    {
  "current": "min",
  "min": {
    "min": true,
    "max": true,
    "only": true
  },
  "max": {
    "min": true,
    "max": true,
    "only": true
  }
}
  </pre>
  <h1>
     Current interval: 
    <span>
      min
    </span>
  </h1>
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</div>
`;

exports[`vue-responsiveness > should work 1`] = `
<div>
  <pre>
    {
  "current": "xs",
  "xs": {
    "min": true,
    "max": true,
    "only": true
  },
  "sm": {
    "min": true,
    "max": true,
    "only": true
  },
  "md": {
    "min": true,
    "max": true,
    "only": true
  },
  "lg": {
    "min": true,
    "max": true,
    "only": true
  },
  "xl": {
    "min": true,
    "max": true,
    "only": true
  },
  "xxl": {
    "min": true,
    "max": true,
    "only": true
  }
}
  </pre>
  <h1>
     Current interval: 
    <span>
      xs
    </span>
  </h1>
  <h3>
    <code>
      sm
    </code>
     and below
  </h3>
  <h3>
    <code>
      md
    </code>
     only
  </h3>
  <h3>
    <code>
      lg
    </code>
     and above
  </h3>
</div>
`;

exports[`vue-responsiveness > should work with custom breakpoints 1`] = `
<div>
  <pre>
    {
  "current": "min",
  "min": {
    "min": true,
    "max": true,
    "only": true
  },
  "max": {
    "min": true,
    "max": true,
    "only": true
  }
}
  </pre>
  <h1>
     Current interval: 
    <span>
      min
    </span>
  </h1>
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</div>
`;
